### ISSUE TITLE
Click on not existed element freezing appium instead of error 500

### ISSUE THREAD

**Issue by isnifer** [appium/appium#8689]
## The problem
Starting from 1.6.5 appium not more send 500 error when I try to click on element that not exists on the screen. 1.6.4 successfully send error 500 that I can handle in my tests.

## Environment

* Appium version (or git revision) that exhibits the issue: 1.6.5
* Last Appium version that did not exhibit the issue (if applicable): 1.6.4
* Desktop OS/version used to run Appium: macOS 10.12.5
* Node.js version (unless using Appium.app|exe): 8.1.0
* Mobile platform/version under test: iOS 10.2
* Real device or emulator/simulator: simulator
* Appium CLI or Appium.app|exe: tipsi-appium-helper

## Link to Appium logs

[1.6.5 Log](https://gist.github.com/isnifer/1b09a91b309ac6c09f34074d1af337d6)
[1.6.4 Log](https://gist.github.com/isnifer/0b629ed637f104222c7bd820a1d5c66d)

## Code To Reproduce Issue [ Good To Have ]
```js
export default async function hideKeyboard() {
  if (this.platform('ios')) {
    const defaultId = this.idFromXPath(`
      //XCUIElementTypeApplication/XCUIElementTypeWindow[5]/XCUIElementTypeOther/
      XCUIElementTypeOther/XCUIElementTypeOther[2]/XCUIElementTypeKeyboard/
      XCUIElementTypeOther/XCUIElementTypeOther/XCUIElementTypeButton[4]
    `)
    const nextId = this.idFromXPath(`
      //XCUIElementTypeApplication/XCUIElementTypeWindow[4]/XCUIElementTypeOther/
      XCUIElementTypeOther/XCUIElementTypeOther[2]/XCUIElementTypeKeyboard/
      XCUIElementTypeOther/XCUIElementTypeOther/XCUIElementTypeButton[4]
    `)

    let isFound = false

    try {
      await this.driver.click(defaultId)
      isFound = true
    } catch (e) {
      await this.driver.click(nextId)
      isFound = true
    }

    if (!isFound) {
      throw new Error('Done button not found')
    }
  } else {
    await this.driver.back()
  }
}
```
## Demo Error (1.6.5)
![1 6 5-error](https://user-images.githubusercontent.com/1788245/27264870-a258fd92-5491-11e7-89ba-58ab03670f97.gif)

## Demo Success (1.6.4)
![1 6 4-success](https://user-images.githubusercontent.com/1788245/27264871-b8e2b9ae-5491-11e7-8b96-525685d08989.gif)

@jlipps @imurchie please, take a look :)

**Comment by mykola-mokhnach** [comment]
Can you try to rewrite your locator to avoid xpath usage? For example

```
const element = await driver.findElement(By.className('XCUIElementTypeKeyboard')).findElement(By.iOSClassChain('XCUIElementTypeOther/XCUIElementTypeOther/XCUIElementTypeButton[4]'))
```

**Comment by mykola-mokhnach** [comment]
Your problem might be https://github.com/facebook/WebDriverAgent/issues/372 (check device logs to make sure) and it happens because of visibility attribute calculation during xpath query evaluation.

**Comment by isnifer** [comment]
@mykola-mokhnach I will try, but I don't want to see workaround for this bug, I hope someone knows how to fix that problem in future releases of appium or any other 3rd party dep

**Comment by JaniJegoroff** [comment]
Just for the reference - I'm always using id's for finding elements and can't reproduce this issue with 1.6.5 nor latest beta.
```
[debug] [MJSONWP] Calling AppiumDriver.findElement() with args: ["id","launch button entercardnumberxxxxx","8bfa021a-3333-4a19-ba70-147d89d82330"]
[debug] [XCUITest] Executing command 'findElement'
[debug] [BaseDriver] Valid locator strategies for this request: xpath, id, name, class name, -ios predicate string, -ios class chain, accessibility id
[debug] [BaseDriver] Waiting up to 0 ms for condition
[debug] [JSONWP Proxy] Proxying [POST /element] to [POST http://localhost:8100/session/C540F5BA-E418-453F-AA3F-E4320D05FAF6/element] with body: {"using":"id","value":"launch button entercardnumberxxxxx"}
[debug] [JSONWP Proxy] Got response with status 200: {"value":{"using":"id","value":"launch button entercardnumberxxxxx","description":"unable to find an element"},"sessionId":"C540F5BA-E418-453F-AA3F-E4320D05FAF6","status":7}
[HTTP] <-- POST /wd/hub/session/8bfa021a-3333-4a19-ba70-147d89d82330/element 500 120 ms - 164
```

**Comment by mykola-mokhnach** [comment]
@isnifer Yep, Apple created it and only Apple knows how to fix. It's like I could change the world, but I doesn't have the source code.

I'm currently working on extending of [class chain locator](https://github.com/facebook/WebDriverAgent/pull/612) (indirect search), so it will be necessary to use xpath in very rare situations.

**Comment by isnifer** [comment]
@mykola-mokhnach Apple didn't do anything between 1.6.4 and 1.6.5, yeah? So, that bug in appium or new version of WDA, I don't know, but I'm sure I'm correct

**Comment by mykola-mokhnach** [comment]
@isnifer the main difference there that visibility attribute has been added back to XML tree representation and exactly visibility calculation causes most of the problems above

**Comment by isnifer** [comment]
@mykola-mokhnach and you're talking that this done by Apple?

**Comment by jlipps** [comment]
@isnifer it sounds like the way visibility calculations are handled in WebDriverAgent changed between the one we included in Appium 1.6.4 and 1.6.5. unfortunately this resulted in a breaking change for you, which is a bad situation. Unfortunately XPath with XCUITest is not stabilized yet, precisely because of these Apple issues that @mykola-mokhnach mentioned. XPath queries are not the recommended way to find elements, partially for this reason. I'm going to close this since I don't think we will reverse the decision to handle visibility in the current way, but future versions of WDA/Appium might fare better with the XPath selectors you're using.

**Comment by isnifer** [comment]
@jlipps no problem, I've started use detox already.
